### DISCUSSION TITLE
fold_inplace initial value is not constexpr

### DISCUSSION THREAD

**Discussion by Life4gal** [foonathan/lexy#97] [Q&A] [answered]
I need to use a custom allocator to manage memory (especially GC), but I found that MSVC refused to compile my code.

```text
error C2131: expression did not evaluate to a constant
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.33.31629\include\xmemory(1062):
note: failure was caused by call of undefined function or one not declared 'constexpr' 
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.33.31629\include\xmemory(1062):
note: see usage of 'StlAllocator<_Newfirst>::allocate' with
[
    _Newfirst=std::_Container_proxy
]
```

Here is a [godbolt](https://godbolt.org/z/fo4bWfjEv) link(GCC/CLang works well)

**Comment by foonathan** [comment, marked as answer]
Instead of passing an initial value, pass a lambda that returns that value: https://godbolt.org/z/xoa1xdbGj

**Reply by Life4gal** [reply]
It solves the problem, but why?

In [Example 2. Parse a list of numbers, but store them in reverse](https://lexy.foonathan.net/reference/callback/fold/#fold_inplace), it also passes a container (initializer_list), but why can it run? I also tried to change it to vector (with std::allocator), and it still works

**Reply by foonathan** [reply]
Without the lambda, it requires that the initial element is constexpr. The other types are, but the container with custom allocator isn't on MSVC.

**Comment by Life4gal** [comment]
![image](https://user-images.githubusercontent.com/52756109/200112008-c7811be2-8b45-4d73-a2da-bbd1ac32310b.png)
Another puzzling point is that once I put them into lambda, it will not compile. :(

**Reply by foonathan** [reply]
You need to immediately invoke the lambda:

`rule = []{ ... };` -> `rule = []{ ... }();`

**Reply by Life4gal** [reply]
Oh, I forgot it!
